src/node_report.h

```diff
@@ -0,0 +1,190 @@
+         << ": "                                                               \
+         << "\"" << value << "\""                                              \
+         << ","                                                                \
+         << "\n";                                                              \
+  } while (0)
+#define ARRAYENTRYNORMAL(jstr, intend, value)                                  \
+  do {                                                                         \
+    ADVANCE(jstr, intend);                                                     \
+    jstr << "\"" << value << "\""                                              \
+         << ","                                                                \
+         << "\n";                                                              \
+  } while (0)
+#define KEYVALUEFINAL(jstr, intend, key, value)                                \
+  do {                                                                         \
+    ADVANCE(jstr, intend);                                                     \
+    jstr << "\"" << key << "\""                                                \
+         << ": "                                                               \
+         << "\"" << value << "\""                                              \
+         << "\n";                                                              \
+  } while (0)
+
+#define ARRAYENTRYFINAL(jstr, intend, value)                                   \
+  do {                                                                         \
+    ADVANCE(jstr, intend);                                                     \
+    jstr << "\"" << value << "\""                                              \
+         << "\n";                                                              \
+  } while (0)
+
+#define ARRAYENDNORMAL(jstr, intend)                                           \
+  do {                                                                         \
+    jstr << "\n";                                                              \
+    ADVANCE(jstr, intend);                                                     \
+    jstr << "],"                                                               \
+         << "\n";                                                              \
+  } while (0)
+
+#define ARRAYENDFINAL(jstr, intend)                                            \
+  do {                                                                         \
+    jstr << "\n";                                                              \
+    ADVANCE(jstr, intend);                                                     \
+    jstr << "]"                                                                \
+         << "\n";                                                              \
+  } while (0)
+
+#define UNREAD                                                                 \
```

Comment:
This should probably take a parameter instead of assuming `out`.

Reply:
yes, moved all these to inline functions, that takes required parameters.